### DISCUSSION TITLE
plot_persistence_diagram depends on non-default latex packages

### DISCUSSION THREAD

**Discussion by mglisse** [GUDHI/gudhi-devel#1210] [General]
See https://github.com/GUDHI/TDA-tutorial/issues/16 for what happens when someone has a basic (instead of full) installation of latex. Not sure what, if anything, we should do about it, but it isn't a nice behavior.

**Comment by VincentRouvreau** [comment]
* The [matplotlib documentation](https://matplotlib.org/3.2.1/tutorials/text/usetex.html) says: "Matplotlib's LaTeX support requires a working LaTeX installation, dvipng (which may be included with your LaTeX installation), and Ghostscript (GPL Ghostscript 9.0 or later is required). The executables for these external dependencies must all be located on your PATH."
* It seems to be the same on [dependencies installation](https://matplotlib.org/3.2.1/users/installing.html#dependencies).
* In this [stack](https://stackoverflow.com/questions/31214214/matplotlib-error-latex-was-not-able-to-process-the-following-string-lp), they say it works if you `sudo apt-get install texlive-latex-extra texlive-fonts-recommended dvipng` but it didn't work for me on a docker based on a Ubuntu 20.04.
* In [third party distributions of matplotlib](https://matplotlib.org/3.2.1/users/installing.html#third-party-distributions-of-matplotlib), they seem to recommend to do `sudo apt-get install python-matplotlib` (or  `sudo apt-get install python3-matplotlib`) and it worked better in my case (and the docker image is smaller)
This debian package is also installing inkscape, ghostscript and dvipng

**Comment by mglisse** [comment]
Would it make sense to do the call to matplotlib that uses latex inside try-except, and in case it fails (in some specific way?) print a warning, disable latex for this session, and rerun the command without the latex option? ISTR that we use latex for esthetic reasons only, but I may remember wrong.

**Comment by VincentRouvreau** [comment]
This would be a nice move !
It seems you need to install texlive-full, I cannot find an exhaustive list... A bit too much just to display persistence diagram...

**Comment by VincentRouvreau** [comment]
Argh, the problem is that the issue appears on `plt.show()`, not before:
```python
import matplotlib 
matplotlib.rcParams['text.usetex'] = True 
import matplotlib.pyplot as plt                                                                      
plt.title(r'\TeX\ ')                                                                                 
plt.show()  # ==> Exception
```
But ok, we can do something on import. Do you prefer a warning on `import gudhi` or on each `gudhi.plot_...` call ? The second one seems better to me.

**Comment by mglisse** [comment]
> the problem is that the issue appears on `plt.show()`

Ah, that's a problem indeed.

> But ok, we can do something on import. Do you prefer a warning on `import gudhi` or on each `gudhi.plot_...` call ? The second one seems better to me.

I don't know what the "something" is, so I can't easily comment.
It doesn't seem useful to have the warning multiple times (and indeed if we use warnings.warn, I think by default it only prints the warning the first time).

**Comment by VincentRouvreau** [comment]
Partially fixed on #357 but matplotlib method does not detect if `type1cm` LaTeX package is installed (required for matplotlib LaTeX).

**Comment by VincentRouvreau** [comment]
This [gist](https://gist.github.com/VincentRouvreau/c101fbce20803ab05071d7fdb005850c) shows how to reproduce the `type1cm` error.

**Comment by mglisse** [comment]
Copied from #368:

I wonder if setting the (global) matplotlib font during a call to those functions is really a good idea. If the user likes some specific font, couldn't they set it themselves once at the beginning of their program? And if they do, it looks like the current code not only overrides their choice for the gudhi functions, it actually overwrites their choice also for future direct calls to matplotlib by the user. So one option would be to remove those settings from the gudhi code, and instead advertise in the doc that we recommend users call matplotlib.pyplot.rc(...) (use latex + CM font) to get prettier plots.
I haven't thought much about it, just putting the idea out there to see what people think. Let me also ping @tlacombe .

**Comment by VincentRouvreau** [comment]
_[UPDATE 06/2025] I moved this issue as a discussion._

Should we keep the default LaTeX behaviour for persistence diagrams/barcode as we are facing issues that doesn't seem to be fixable ?
Or Should we let it to the user responsibility ? (as matplotlib is recently doing)
